Bug: test_voq_counter: Credit-WD-Del counter does not increment on DNX/Q3D single-ASIC platforms

### Is it platform specific

broadcom

### Importance or Severity

Medium

### Description of the bug

The `tests/voq/test_voq_counter.py::test_voq_drop_counter` test checks the
Credit-WD-Del counter but does not observe any counter increments on
DNX/Q3D single-ASIC platforms, even when the Credit Watchdog feature is
working correctly.

The current test logic makes several assumptions that are not valid on
these platforms:

* It assumes the presence of fabric/SFI ports and uses
  `port enable sfi <true/false>` to trigger Credit Watchdog deletes.
  DNX/Q3D single-ASIC platforms have no fabric ports, so this has no
  effect on Credit Watchdog.
* It parses Credit-WD-Del counters from VOQ0 for Ethernet-IB links.
  On DNX/Q3D single-ASIC platforms there may be no Ethernet-IB links,
  and traffic accumulates in a different VOQ index (e.g. VOQ7), so the
  test is looking at the wrong counters.
* It always uses `bcmcmd -n`, which is only applicable to multi-ASIC
  systems, not single-ASIC DNX/Q3D.

As a result, the test either fails or reports zero Credit-WD-Del packets
on DNX/Q3D single-ASIC platforms, even though Credit Watchdog is actually
deleting packets as expected.

### Steps to Reproduce

1. Set up a DNX/Q3D-based single-ASIC DUT with a VOQ topology supported
   by `sonic-mgmt`.
2. Deploy a SONiC image where Credit Watchdog is enabled and operational
   on the device.
3. From the `sonic-mgmt` test environment, run test voq/test_voq_counter.py::test_voq_drop_counter
4. Observe the behavior of the `Credit-WD-Del` counters that the test
   parses, and the final test result.



### Actual Behavior and Expected Behavior

**Actual behavior**

On DNX/Q3D single-ASIC platforms:

* Running `tests/voq/test_voq_counter.py::test_voq_drop_counter` does
  not cause the parsed `Credit-WD-Del` counters to increment, because:
  * `port enable sfi <true/false>` is a no-op (no fabric ports).
  * The test parses counters from VOQ0 for Ethernet-IB links, which do
    not exist on this platform and are not where traffic accumulates.
  * The test always uses `bcmcmd -n`, which is not applicable to
    single-ASIC systems.

Depending on environment and expectations, this can cause the test to
fail or to report zero Credit-WD-Del packets even when Credit Watchdog
is active and deleting packets.

**Expected behavior**

* The test should induce Credit Watchdog deletes in a way that is valid
  for DNX/Q3D single-ASIC platforms (e.g. by using
  `setreg SCH_SCHEDULER_CONFIGURATION_REGISTER DISABLE_FABRIC_MSGS` to
  starve the central scheduler).
* It should parse Credit-WD-Del counters from the VOQ indices where
  traffic actually queues on these platforms (e.g. VOQ7), rather than
  assuming VOQ0 for Ethernet-IB.
* It should avoid `bcmcmd -n` on single-ASIC systems, while preserving
  the existing multi-ASIC path.

With these adjustments, the test should:
* increment the relevant `Credit-WD-Del` counters on DNX/Q3D single-ASIC
  platforms when Credit Watchdog deletes occur, and
* pass when the feature is working correctly.

### Relevant log output

```shell
def test_voq_queue_counter(duthosts, enum_rand_one_per_hwsku_frontend_hostname):
        """
        This test implicitly verifies that queue counters --voq (i.e. Credit-WD-Del/pkts)
        are working as expected by disabling the fabric ports
        """
        duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
        bcm_changes = False
        # Ensure the device is a Broadcom device
        pytest_require((duthost.facts.get('platform_asic') == "broadcom-dnx"),
                       "The Test Case is only supported on Broadcom-dnx ASIC")
    
        cmd_bcmcmd_false = "'port enable sfi false'"
        cmd_bcmcmd_true = "'port enable sfi true'"
        cmd = "show queue counters --voq --nonzero| grep -i '{}' |grep -i '{}' |awk '{{print $7}}'".format("Ethernet-IB",
                                                                                                           "VOQ0")
        try:
            bcm_changes = True
            for asic in duthost.asics:
                bcmcmd = "bcmcmd {} ".format("-n " + str(asic.asic_index)) + cmd_bcmcmd_false
                res = duthost.shell(bcmcmd, module_ignore_errors=True)
                if not res["stderr"] == "polling socket timeout: Success" and res["failed"]:
                    pytest.fail("BCMCMD Failed")
    
            def queue_counter_assertion():
                out = duthost.shell(cmd)['stdout'].split('\n')
                integers = [int(item.replace(',', '')) for item in out if item.replace(',', '').strip().isdigit()]
                return any(num > 0 for num in integers)
    
>           pytest_assert(wait_until(300, 0, 0, queue_counter_assertion),
                          "Credit-WD-Del/pkts is not increasing "
                          "Ref: https://github.com/sonic-net/sonic-buildimage/issues/21098")
E                         Failed: Credit-WD-Del/pkts is not increasing Ref: https://github.com/sonic-net/sonic-buildimage/issues/21098
```

### Output of `show version`

```shell

```

### Attach files (if any)

_No response_

--- a/tests/voq/test_voq_counter.py
+++ b/tests/voq/test_voq_counter.py
@@ -2,6 +2,7 @@ import logging
 import pytest
 from tests.common.helpers.assertions import pytest_assert, pytest_require
 from tests.common.utilities import wait_until
+from tests.common.gu_utils import get_asic_name
 
 
 logger = logging.getLogger(__name__)
@@ -26,6 +27,7 @@ def test_voq_queue_counter(duthosts, enum_rand_one_per_hwsku_frontend_hostname):
     """
     This test implicitly verifies that queue counters --voq (i.e. Credit-WD-Del/pkts)
     are working as expected by disabling the fabric ports
+    For Q3D (single-ASIC), instead disable fabric messages via register setting.
     """
     duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
     bcm_changes = False
@@ -33,30 +35,62 @@ def test_voq_queue_counter(duthosts, enum_rand_one_per_hwsku_frontend_hostname):
     pytest_require((duthost.facts.get('platform_asic') == "broadcom-dnx"),
                    "The Test Case is only supported on Broadcom-dnx ASIC")
 
-    cmd_bcmcmd_false = "'port enable sfi false'"
-    cmd_bcmcmd_true = "'port enable sfi true'"
-    cmd = "show queue counters --voq --nonzero| grep -i '{}' |grep -i '{}' |awk '{{print $7}}'".format("Ethernet-IB",
-                                                                                                       "VOQ0")
-    try:
-        bcm_changes = True
-        for asic in duthost.asics:
-            bcmcmd = "bcmcmd {} ".format("-n " + str(asic.asic_index)) + cmd_bcmcmd_false
+    duthost.shell("sonic-clear queuecounters")
+
+    asic_name = get_asic_name(duthost).lower()
+    is_q3d_single_asic = ("q3d" in asic_name) and (not duthost.is_multi_asic)
+
+    if is_q3d_single_asic:
+        cmd_bcmcmd = "setreg SCH_SCHEDULER_CONFIGURATION_REGISTER DISABLE_FABRIC_MSGS"
+        cmd = "show queue counters --voq --nonzero | grep -i 'VOQ7' | awk '{print $7}'"
+
+        try:
+            bcm_changes = True
+            bcmcmd = f"bcmcmd '{cmd_bcmcmd}'=1"
             res = duthost.shell(bcmcmd, module_ignore_errors=True)
             if not res["stderr"] == "polling socket timeout: Success" and res["failed"]:
-                pytest.fail("BCMCMD Failed")
+                pytest.fail("BCMCMD Failed to disable fabric messages")
 
-        def queue_counter_assertion():
-            out = duthost.shell(cmd)['stdout'].split('\n')
-            integers = [int(item.replace(',', '')) for item in out if item.replace(',', '').strip().isdigit()]
-            return any(num > 0 for num in integers)
+            def queue_counter_assertion():
+                out = duthost.shell(cmd)["stdout"].split("\n")
+                integers = [int(item.replace(",", "")) for item in out if item.replace(",", "").strip().isdigit()]
+                return any(num > 0 for num in integers)
 
-        pytest_assert(wait_until(300, 0, 0, queue_counter_assertion),
-                      "Credit-WD-Del/pkts is not increasing "
-                      "Ref: https://github.com/sonic-net/sonic-buildimage/issues/21098")
-    finally:
-        if bcm_changes:
+            pytest_assert(
+                wait_until(300, 5, 0, queue_counter_assertion),
+                "Credit-WD-Del/pkts counter did not increment. "
+                "Ref: https://github.com/sonic-net/sonic-buildimage/issues/21098",
+            )
+        finally:
+            if bcm_changes:
+                bcmcmd = f"bcmcmd '{cmd_bcmcmd}'=0"
+                res = duthost.shell(bcmcmd, module_ignore_errors=True)
+                if not res["stderr"] == "polling socket timeout: Success" and res["failed"]:
+                    pytest.fail("BCMCMD Failed to re-enable fabric messages")
+    else:
+        cmd_bcmcmd_false = "'port enable sfi false'"
+        cmd_bcmcmd_true = "'port enable sfi true'"
+        cmd = "show queue counters --voq --nonzero| grep -i 'Ethernet-IB' |grep -i 'VOQ0' |awk '{{print $7}}'"
+        try:
+            bcm_changes = True
             for asic in duthost.asics:
-                cmd = "bcmcmd {} ".format("-n " + str(asic.asic_index)) + cmd_bcmcmd_true
-                res = duthost.shell(cmd, module_ignore_errors=True)
+                bcmcmd = "bcmcmd {} ".format("-n " + str(asic.asic_index)) + cmd_bcmcmd_false
+                res = duthost.shell(bcmcmd, module_ignore_errors=True)
                 if not res["stderr"] == "polling socket timeout: Success" and res["failed"]:
                     pytest.fail("BCMCMD Failed")
+
+            def queue_counter_assertion():
+                out = duthost.shell(cmd)['stdout'].split('\n')
+                integers = [int(item.replace(',', '')) for item in out if item.replace(',', '').strip().isdigit()]
+                return any(num > 0 for num in integers)
+
+            pytest_assert(wait_until(300, 0, 0, queue_counter_assertion),
+                          "Credit-WD-Del/pkts is not increasing "
+                          "Ref: https://github.com/sonic-net/sonic-buildimage/issues/21098")
+        finally:
+            if bcm_changes:
+                for asic in duthost.asics:
+                    cmd = "bcmcmd {} ".format("-n " + str(asic.asic_index)) + cmd_bcmcmd_true
+                    res = duthost.shell(cmd, module_ignore_errors=True)
+                    if not res["stderr"] == "polling socket timeout: Success" and res["failed"]:
+                        pytest.fail("BCMCMD Failed")

